Trusted Devices page

flyinghermit opened the pull request:
- Adds Trusted Device page to help feature discovery in community edition. Page does not implement any functionality except for explaining Device Trust feature and showing a CTA button. 
- Updates `ButtonLockedFeature`'s  href to point to `upgrade-community` in oss and `upgrade-team` in team

The design to show the page as a "locked feature" is borrowed from the original implementation https://github.com/gravitational/teleport.e/pull/1200 which was recently removed from `e` as Team plan now offers Device Trust.

flyinghermit committed:
add device trust page

sshahcodes commented:
page screenshot:

<img width="1403" alt="Screenshot 2023-08-09 at 7 29 17 PM" src="https://github.com/gravitational/teleport/assets/10177602/c4f6c8a2-7cbb-4630-9f11-e91ac2a1f237">

flyinghermit committed:
move 'view' before link

zmb3 commented on web/packages/teleport/src/DeviceTrust/DeviceTrustLocked.tsx:
```diff
@@ -0,0 +1,170 @@
+import React from 'react';
```
Don't forget license header.

sshahcodes replied:
Oh yes! added in https://github.com/gravitational/teleport/pull/30249/commits/eb4015be958e231bf29722bb43d44c58dd5ea34f

zmb3 commented on web/packages/teleport/src/DeviceTrust/DeviceTrustLocked.tsx:
```diff
@@ -0,0 +1,170 @@
+          >
+            <IconCircle Icon={LockIcon} size={64} />
+          </Box>
+          <Text fontSize="2" textAlign="center">
```
I think the centered text looks awkward here. What say you, @roraback?

sshahcodes replied:
Adjusted text alignment to `justify` and reduced width to `600px` so that it looks uniform in https://github.com/gravitational/teleport/pull/30249/commits/a6ac29fbff1c4ff851779c6ab67c42c0787e05b1

<img width="1026" alt="Screenshot 2023-08-15 at 10 04 59 AM" src="https://github.com/gravitational/teleport/assets/10177602/bfa77deb-32c4-47f6-8f20-4968535330a7">

zmb3 commented:
The other thing I noticed is the page is scrollable, even though its contents fit on screen.

flyinghermit committed:
add license header

flyinghermit committed:
add static height and adjust css to prevent vertial scrolling

sshahcodes commented:
> The other thing I noticed is the page is scrollable, even though its contents fit on screen.

Thanks for pointing out. I did not notice in my default window setup but I could verify by reducing my browser's viewport. I have updated css in https://github.com/gravitational/teleport/pull/30249/commits/9eca83a13be51efd86923f06747de23f1722f7fb which should fix that 🤞

flyinghermit committed:
udpate ButtonLockedFeature's  href to point to upgrade-community in oss and upgrade-team in team

flyinghermit committed:
update ButtonLockedFeature test to account for team vs oss links

flyinghermit committed:
fix linter

sshahcodes commented:
Friendly ping for review

ryanclark commented on web/packages/teleport/src/DeviceTrust/DeviceTrustLocked.tsx:
```diff
@@ -0,0 +1,179 @@
+import React from 'react';
+import styled, { useTheme } from 'styled-components';
+
+import Box from 'design/Box';
+
+import Text from 'design/Text';
+
+import Flex from 'design/Flex';
+
+import Link from 'design/Link';
```
Can you group these together?

ryanclark commented on web/packages/teleport/src/DeviceTrust/DeviceTrustLocked.tsx:
```diff
@@ -0,0 +1,179 @@
+  pageSize = 20,
+  fetchStatus = '',
+  fetchData,
+}: any) => {
```
Can you add types here?

ryanclark commented on web/packages/teleport/src/DeviceTrust/DeviceTrustLocked.tsx:
```diff
@@ -0,0 +1,179 @@
+  );
+};
+
+const IconCell = ({ osType }) => {
```
types here too

ryanclark commented on web/packages/teleport/src/components/ButtonLockedFeature/ButtonLockedFeature.tsx:
```diff
@@ -50,7 +53,9 @@ export function ButtonLockedFeature({
       as="a"
       target="blank"
-      href={`${SALES_URL}?${getParams(version, isEnterprise, event)}`}
+      href={`${
+        cfg.isUsageBasedBilling ? UPGRADE_TEAM_URL : UPGRADE_COMMUNITY_URL
+      }?${getParams(version, isEnterprise, event)}`}
```
I think defining this above and referencing the variable would look a bit neater

```typescript
const baseUrl = cfg.isUsageBasedBilling ? UPGRADE_TEAM_URL : UPGRADE_COMMUNITY_URL;
const upgradeUrl = `${baseUrl}?${getParams(version, isEnterprise, event)}`;
```

ryanclark reviewed (approved):
few nits

flyinghermit committed:
Merge branch 'master' into sshah/trusted-device-cta-page-oss

flyinghermit committed:
apply review suggestions:
- add types: this is now exported to be used in e as well
- group design imports
- create vars for upgrade urls

flyinghermit committed:
remove export IconCell

avatus commented on web/packages/teleport/src/DeviceTrust/DeviceTrustLocked.tsx:
```diff
@@ -0,0 +1,178 @@
+export function DeviceTrustLocked() {
+  const theme = useTheme();
+
+  function generateFakeItems(count) {
+    const items = [];
+    const osType = ['Windows', 'Linux', 'macOS'];
+
+    for (let i = 0; i < count; i++) {
+      items.push({
+        id: `id-${i}`,
+        assetTag: `CLFBDAACC${i}`,
+        enrollStatus: `enroll-status-${i}`,
+        osType: osType[i % osType.length],
+      });
+    }
+
+    return items;
+  }
```
very small nit: this could be moved outside of the function so it doesn't rerun every render... although we really only render this once so it doesn't matter.

sshahcodes replied:
moved outside in https://github.com/gravitational/teleport/pull/30249/commits/b5078ca2d60839ffef0921e717518035e74e26ca so it does not breaks the recommended pattern

ibeckermayer commented on web/packages/teleport/src/DeviceTrust/DeviceTrustLocked.tsx:
```diff
@@ -0,0 +1,178 @@
+      fetching={{ onFetchMore: fetchData, fetchStatus }}
+    />
+  );
+};
+
+const IconCell = ({ osType }: { osType: TrustedDeviceOSType }) => {
+  let icon;
+  switch (osType) {
+    case 'Windows':
+      icon = <Windows size="small" mr={1} />;
+      break;
+    case 'Linux':
+      icon = <Linux size="small" mr={1} />;
+      break;
+    default:
+      icon = <Apple size="small" mr={1} />;
+  }
+  return (
+    <Cell align="left" style={{ display: 'flex' }}>
+      {icon} {osType}
+    </Cell>
+  );
+};
+
+const StyledMessageContainer = styled(Flex)`
+  position: relative;
+  top: 30%;
+  left: 50%;
+  transform: translate(-50%, -50%);
+  background-color: ${({ theme }) => theme.colors.levels.elevated};
+  flex-direction: column;
+  justify-content: center;
+  align-items: center;
+  padding: 24px;
+  gap: 24px;
+  width: 650px;
+  box-shadow: 0 5px 5px -3px rgba(0, 0, 0, 0.2),
+    0 8px 10px 1px rgba(0, 0, 0, 0.14), 0 3px 14px 2px rgba(0, 0, 0, 0.12);
+  border-radius: 8px;
+`;
```
Have we considered the possibility of moving these from `e/` to `shared/`, and then just injecting the fake data into this component, rather than having them duplicated?

sshahcodes replied:
I gave it a thought but then decided to do it as its done currently because:

- The table required for `e` itself is holding up to another TODO (we want to use range based server side query but it is not currently implemented in Device Trust service, but can be updated in future). It still may not be a problem but then changes to the `e` table may require to update this one, which IMO brings unnecessary future work for the "fake list table". 
- We aren't sure if the design and CTA for this page will remain the same. The effort may not worth it if we want to update the design in future.

I would ideally move these into `shared/` only if we are to use actual functionality both in `oss` and `e`.

ibeckermayer reviewed (commented):
Realize now that I've been removed from the reviewers list, but I figure this is worth consideration anyhow:

flyinghermit committed:
move generateFakeItems out from DeviceTrustLocked

flyinghermit committed:
adjust CTA text alignment
